handler/src/test/java/io/netty/handler/stream/ChunkedWriteHandlerTest.java

```diff
@@ -227,6 +228,207 @@ public long progress() {
+            public boolean isEndOfInput() throws Exception {
+                return done;
+            }
+
+            @Override
+            public void close() throws Exception {
+                buffer.release();
+                closeWasCalled.set(true);
+            }
+
+            @Deprecated
+            @Override
+            public ByteBuf readChunk(ChannelHandlerContext ctx) throws Exception {
+                return readChunk(ctx.alloc());
+            }
+
+            @Override
+            public ByteBuf readChunk(ByteBufAllocator allocator) throws Exception {
+                if (done) {
+                    return null;
+                }
+                done = true;
+                return buffer.retainedDuplicate();
+            }
+
+            @Override
+            public long length() {
+                return -1;
+            }
+
+            @Override
+            public long progress() {
+                return 1;
+            }
+        };
+
+        EmbeddedChannel ch = new EmbeddedChannel(new ChunkedWriteHandler());
+
+        // write 3 messages and close channel before flushing
+        ChannelFuture r1 = ch.write(new ChunkedFile(TMP));
+        ChannelFuture r2 = ch.write(new ChunkedNioFile(TMP));
+        ch.write(notifiableInput);
+        ch.close().awaitUninterruptibly();
+
+        assertFalse(r1.isSuccess());
+        assertFalse(r2.isSuccess());
+        assertTrue(closeWasCalled.get());
+    }
+
+    // See https://github.com/netty/netty/issues/8700.
+    @Test
+    public void testStopConsumingChunksWhenFailed() {
+        final ByteBuf buffer = Unpooled.copiedBuffer("Test", CharsetUtil.ISO_8859_1);
+        final AtomicInteger chunks = new AtomicInteger(0);
+
+        ChunkedInput<ByteBuf> nonClosableInput = new ChunkedInput<ByteBuf>() {
+            @Override
+            public boolean isEndOfInput() throws Exception {
+                return chunks.get() >= 5;
+            }
+
+            @Override
+            public void close() throws Exception {
+                // no-op
+            }
+
+            @Deprecated
+            @Override
+            public ByteBuf readChunk(ChannelHandlerContext ctx) throws Exception {
+                return readChunk(ctx.alloc());
+            }
+
+            @Override
+            public ByteBuf readChunk(ByteBufAllocator allocator) throws Exception {
+                chunks.incrementAndGet();
+                return buffer.retainedDuplicate();
+            }
+
+            @Override
+            public long length() {
+                return -1;
+            }
+
+            @Override
+            public long progress() {
+                return 1;
+            }
+        };
+
+        ChannelOutboundHandlerAdapter noOpWrites = new ChannelOutboundHandlerAdapter() {
+            @Override
+            public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
+                promise.tryFailure(new RuntimeException());
```

Comment:
You also need to call `ReferenceCountUtil.release(msg)` before failing the promise.